server/src/main/java/io/deephaven/server/object/ObjectServiceGrpcImpl.java

```diff
@@ -197,4 +309,47 @@ public Optional<String> type() {
             return Optional.ofNullable(type);
         }
     }
+
+    private final class PluginMessageSender implements ObjectType.MessageStream {
+        private final StreamObserver<StreamResponse> responseObserver;
+        private final SessionState sessionState;
+
+        public PluginMessageSender(StreamObserver<StreamResponse> responseObserver, SessionState sessionState) {
+            this.responseObserver = responseObserver;
+            this.sessionState = sessionState;
+        }
+
+        @Override
+        public void onMessage(ByteBuffer message, Object[] references) {
+            ExportCollector exportCollector = new ExportCollector(sessionState);
+
+            Data.Builder payload = Data.newBuilder().setPayload(ByteString.copyFrom(message));
+
+            try {
+                for (Object reference : references) {
+                    // allowUnknownType=true so that the caller can let the client handle complex lifecycles
+                    // forceNew=true to explicitly state that the plugin is responsible for handling dups, both
+                    // within a single payload and across the stream
+                    exportCollector.reference(reference, true, true);
+                }
+                for (ReferenceImpl ref : exportCollector.refs()) {
+                    payload.addTypedExportIds(ref.typedTicket());
+                }
+            } catch (RuntimeException | Error t) {
+                exportCollector.cleanup(t);
+                throw t;
+            } catch (Throwable t) {
+                exportCollector.cleanup(t);
+                throw new RuntimeException(t);
+            }
```

Comment:
This reflects the original code - I branched it to avoid wrapping when not necessary, but still handle any sneaky-throw (as seemed the original intent?) from your code, see line ~112 of the "before" diff.

Technically should handle interrupt exception separately, but that would be a bad one to sneak-throw...